crates/sui-core/src/authority.rs

```diff
@@ -3432,6 +3435,46 @@ fn calculate_checkpoint_numbers(
     }
 }
 
+/// In dry run and dev inspect, you might load a dynamic field that is actually too new for the
+/// transaction. Ideally, we would want to load the "correct" dynamic fields, but as that is
+/// not easily determined, we instead set the sequence number of the gas objects to MAX - 1, to
+/// ensure that the lamport version for the transaction is MAX, and as such, a valid lamport
+/// version for any object used in the transaction (preventing internal assertions or
+/// invariant violations from being triggered)
+fn advance_gas_object_version_for_mock_transaction(
+    gas_object_refs: &mut Vec<(ObjectID, SequenceNumber, ObjectDigest)>,
+    input_objects: &mut InputObjects,
+) {
+    const FAKE_SEQ: SequenceNumber = SequenceNumber::from_u64(SequenceNumber::MAX.value() - 1);
+    let mut remapped = HashSet::new();
+    for (id, seq, _digest) in gas_object_refs {
+        remapped.insert(*id);
+        *seq = FAKE_SEQ
+    }
+    let objs = std::mem::replace(input_objects, /* dummy */ InputObjects::new(vec![]));
+    let objs = objs.into_objects();
```

Comment:
Seeing as you're going to this trouble, would it be easier to accept `input_objects` by value and return the new one?